rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
+}
+```
+
+Currently, a few components (at least `Accordion`, `AccordionHeader`, `AccordionItem`) are defining a value for React Context on its state to use it in `render*` functions:
+
+```tsx
+// ⚠ simplified example
+
+function useBazState(props) {
+  const { inline, size } = props;
+  const [open, setOpen] = React.useState();
+
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+  const state = {
+    inline,
+    size,
+    open,
+    // 👇 "contextValue" is a part of component's state
+    contextValue,
+  };
+
+  return state;
+}
+
+function renderBaz(state) {
+  // 👇 "value" is memoized, consumer components will not rerender without need
+  return (
+    <SampleContext.Provider value={state.contextValue}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+## Problem statement
+
+The problem comes from multiple places:
+
+- (_design issue_) `contextValue` is not a part of `state`, it's useless from component's perspective
+- (_usage issue_) in a case of composition state mutations are not intuitive
+
+The last item is a really serious problem as customers should be aware about our implementation details 🕵️‍
+
+```tsx
+// 🛠 In this example I would like to override `size` based on a condition
+function useSuperBazState(props) {
+  const state = useBazState(props);
+
+  // ❌ is broken because `state.contextValue.size` remains the same
+  if (props.reallyBig) {
+    state.size = 'xxl';
+  }
+
+  // ❌ is broken because `state.contextValue` is a stable object and will not trigger
+  //     a rerender
+  if (props.reallyBig) {
+    state.size = 'xxl';
+    state.contextValue.size = 'xxl';
+  }
+
+  // ✅ only this solution will work as expected
+  state.size = 'xxl';
+  state.contextValue = React.useMemo(
+    () => ({
+      ...contextValue,
+      ...(props.reallyBig && { size: 'xxl' }),
+    }),
+    [state.contextValue, props.reallyBig],
+  );
+}
+```
+
+Ideally, we can simply remove `contextValue` from `state` and move it into `render*` function. And it's already done in few components:
+
+```tsx
+function renderBaz(state) {
+  const { open, inline } = state;
+
+  // 💥 this violates rules of hooks: hooks should be called inside components or other hooks
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  return (
+    <SampleContext.Provider value={contextValue}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+The problem is this approach violates [Rules of Hooks](https://reactjs.org/docs/hooks-rules.html) as we are calling `React.useMemo()` inside a function. Actually, it's the reason why `contextValue` is being declared in `use*State()` hooks.
+
+## Solutions
+
+It's clear that we would like to avoid having values for React Context in `state` objects and we don't want to violate Rules of Hooks 🤔
+
+_There is no recomended option yet._
+
+### Option 1: Make `render` functions hooks
+
+We can transform existing render functions to be React hooks:
```

Comment:
I think we should also point out that we will now return a render function from this hook